util.c

```diff
@@ -1234,3 +1235,58 @@ int set_cpu_affinity(cpuset_t *set)
 }
 #endif
 #endif /* HAVE_CPUSET_T */
+
+void add_cookie_secret(struct nsd* nsd, uint8_t* secret)
+{
+	/* New cookie secret becomes the staging secret (position 1)
+	 * unless there is no active cookie yet, then it becomes the active
+	 * secret.  If the NSD_COOKIE_HISTORY_SIZE > 2 then all staging cookies
+	 * are moved one position down.
+	 */
+	if(nsd->cookie_count == 0) {
+		memcpy( nsd->cookie_secrets->cookie_secret
+		       , secret, NSD_COOKIE_SECRET_SIZE);
+		nsd->cookie_count = 1;
+		return;
+	}
+#if NSD_COOKIE_HISTORY_SIZE > 2
+	memmove( &nsd->cookie_secrets[1], &nsd->cookie_secrets[2]
+	       , sizeof(struct cookie_secret) * (NSD_COOKIE_HISTORY_SIZE - 2));
+#endif
+	memcpy( nsd->cookie_secrets[1].cookie_secret
+	      , secret, NSD_COOKIE_SECRET_SIZE);
+	nsd->cookie_count = nsd->cookie_count     < NSD_COOKIE_HISTORY_SIZE
+	                  ? nsd->cookie_count + 1 : NSD_COOKIE_HISTORY_SIZE;
+	memset(secret, 0, NSD_COOKIE_SECRET_SIZE);
+}
+
+void activate_cookie_secret(struct nsd* nsd)
+{
+	uint8_t active_secret[NSD_COOKIE_SECRET_SIZE];
+	/* The staging secret becomes the active secret.
+	 * The active secret becomes a staging secret.
+	 * If the NSD_COOKIE_HISTORY_SIZE > 2 then all stagin secrets are moved
+	 * one position up and the previously active secret becomes the last
+	 * staging secret.
+	 */
+	if(nsd->cookie_count < 2)
+		return;
+	memcpy( active_secret, nsd->cookie_secrets[0].cookie_secret
+	      , NSD_COOKIE_SECRET_SIZE);
+	memmove( &nsd->cookie_secrets[0], &nsd->cookie_secrets[1]
+	       , sizeof(struct cookie_secret) * (NSD_COOKIE_HISTORY_SIZE - 1));
+	memcpy( nsd->cookie_secrets[nsd->cookie_count - 1].cookie_secret
+	      , active_secret, NSD_COOKIE_SECRET_SIZE);
+	memset(active_secret, 0, NSD_COOKIE_SECRET_SIZE);
```

Comment:
These kinds of zero memsets get removed by the compiler. Use `explicit_bzero` to make sure it does not get removed. Also for the other memset operations on the cookies.